### PR TITLE
Fix JSON serialiation of `ValueEnum`s

### LINKED ISSUE #914: Bug: The `.as_dict()` method on `CrystalSystem` is not Monty compatible
### Problem

```python
from emmet.core.symmetry import CrystalSystem
sym = CrystalSystem("Triclinic")
sym.as_dict()
```

This returns the string "Triclinic" rather than a dictionary. This would normally be fine, but the fact that it's a string means that doing `dumpfn` doesn't work on a data structure containing a `CrystalSystem` object.

```python
from monty.serialization import dumpfn

dumpfn(sym, "test.json")
```

Traceback:
```python
------------------------------------------------------------------------
TypeError                              Traceback (most recent call last)
Cell In[47], [line 1](vscode-notebook-cell:?execution_count=47&line=1)
----> [1](vscode-notebook-cell:?execution_count=47&line=1) dumpfn(CrystalSystem("Triclinic"),"test.json")

File [c:\Users\asros\miniconda\envs\zeolites\lib\site-packages\monty\serialization.py:123](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:123), in dumpfn(obj, fn, fmt, *args, **kwargs)
    [121](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:121)     if "cls" not in kwargs:
    [122](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:122)         kwargs["cls"] = MontyEncoder
--> [123](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:123)     fp.write(json.dumps(obj, *args, **kwargs))
    [124](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:124) else:
    [125](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/serialization.py:125)     raise TypeError(f"Invalid format: {fmt}")

File [c:\Users\asros\miniconda\envs\zeolites\lib\json\__init__.py:238](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:238), in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    [232](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:232) if cls is None:
    [233](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:233)     cls = JSONEncoder
    [234](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:234) return cls(
    [235](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:235)     skipkeys=skipkeys, ensure_ascii=ensure_ascii,
    [236](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:236)     check_circular=check_circular, allow_nan=allow_nan, indent=indent,
    [237](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:237)     separators=separators, default=default, sort_keys=sort_keys,
--> [238](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/__init__.py:238)     **kw).encode(obj)

File [c:\Users\asros\miniconda\envs\zeolites\lib\json\encoder.py:199](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:199), in JSONEncoder.encode(self, o)
    [195](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:195)         return encode_basestring(o)
    [196](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:196) # This doesn't pass the iterator directly to ''.join() because the
    [197](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:197) # exceptions aren't as detailed.  The list call should be roughly
    [198](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:198) # equivalent to the PySequence_Fast that ''.join() would do.
--> [199](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:199) chunks = self.iterencode(o, _one_shot=True)
    [200](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:200) if not isinstance(chunks, (list, tuple)):
    [201](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:201)     chunks = list(chunks)

File [c:\Users\asros\miniconda\envs\zeolites\lib\json\encoder.py:257](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:257), in JSONEncoder.iterencode(self, o, _one_shot)
    [252](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:252) else:
    [253](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:253)     _iterencode = _make_iterencode(
    [254](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:254)         markers, self.default, _encoder, self.indent, floatstr,
    [255](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:255)         self.key_separator, self.item_separator, self.sort_keys,
    [256](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:256)         self.skipkeys, _one_shot)
--> [257](file:///C:/Users/asros/miniconda/envs/zeolites/lib/json/encoder.py:257) return _iterencode(o, 0)

File [c:\Users\asros\miniconda\envs\zeolites\lib\site-packages\monty\json.py:394](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:394), in MontyEncoder.default(self, o)
    [391](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:391)     d = o.as_dict()
    [393](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:393) if "@module" not in d:
--> [394](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:394)     d["@module"] = str(o.__class__.__module__)
    [395](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:395) if "@class" not in d:
    [396](file:///C:/Users/asros/miniconda/envs/zeolites/lib/site-packages/monty/json.py:396)     d["@class"] = str(o.__class__.__name__)

TypeError: 'str' object does not support item assignment
```

### Proposed Solution

Modify the `.as_dict()` method?

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Andrew-S-Rosen**
Closes #914.

```python
from monty.serialization import dumpfn
from emmet.core.symmetry import CrystalSystem

sym = CrystalSystem("Triclinic")

dumpfn(sym, "test.json")
```

This now works.

### CODE REVIEW

**Comment by Andrew-S-Rosen** [comment]
Tagging @munrojm. Fixed it, I think! If you wouldn't mind minting a new version after this is merged, that would be a huge help.

Thank you @jmmshn for your suggestion. It worked like a charm! Not sure what I was doing wrong before.

**Comment by jmmshn** [comment]
I will never end the crusade against custom `as_dict` / `from_dict`

**Comment by esoteric-ephemera** [comment]
Hey I think this broke atomate2, see this [atomate2 issue](https://github.com/materialsproject/atomate2/issues/781) which I've reproduced with emmet>=0.78.0, and couldn't reproduce with emmet<=0.77.1

**Comment by Andrew-S-Rosen** [comment]
It's probably better to solve the root cause of the problem in atomate2 (or monty, if relevant) rather than patch back in a custom `.as_dict()` method that also doesn't serialize to a dictionary but rather a string.

**Comment by utf** [comment]
Maybe I'm missing something, but this PR changed the behaviour of these objects. That doesn't seem like an atomate2 issue as we were using these objects as intended.

**Comment by esoteric-ephemera** [comment]
There's a `jsanitize` call in atomate2 with `strict=True` passed that would require `VaspObject` to have an `as_dict` method, which was removed here

**Comment by utf** [comment]
The point is that we want to serialise these objects as strings. That's why we were using ValueEnum. 

I'd ask if you want to have enums that serialise as dicts then please make another object!

**Comment by Andrew-S-Rosen** [comment]
I guess there are a few things to unpack here.

In emmet, and therefore in atomate2 and other packages, there are a few objects inheriting from `ValueEnum`s that break serialization. The example I mentioned in #944 is one example. When the `CrystalSystem` is dumped (e.g. deposited in a MongoDB), you can't reload it. That's because the custom `.as_dict()` method returning a `str` cannot reversibly generate back the parent `CrystalSystem`. This breaks roundtrip `dumpfn`/`loadfn` and is a problem.

The `ValueEnum` in emmet is, admittedly, advertised as a way to serialize the object to `str`. So, I can see the argument there for retaining the `.as_dict()` method to be a `str` return. However, the method is called `.as_dict()`. In my opinion, we should not have a method like `.as_dict()` returning something that is not a dictionary. This has downstream negative effects, as mentioned in the example above, and is arguably a hack.

I suppose the question to ask is there a better way to achieve the desired `str` serialization before without relying on such a hacky/fake `.as_dict()` method? I would argue the answer is yes --- inherit from `Enum` and take advantage of monty's `enum_values=True` [flag](https://github.com/materialsvirtuallab/monty/blob/4520972e78985394b3ef02d0b8ffb1873bc27a61/monty/json.py#L656).

**Comment by utf** [comment]
I think this PR and issue #914 are misunderstanding what `ValueEnum` is for. There many objects in Python that cannot be serialised as json by themselves. For example, `str`, `int`, `float`, `list` etc. Instead, they have to be serialised as part of a dict.  The fact that this does not work

```python
from monty.serialization import dumpfn
from emmet.core.symmetry import CrystalSystem

sym = CrystalSystem("Triclinic")
dumpfn(sym, "test.json")
```
is not an issue in itself. This is equivalent to saying the following means that string serialisation is broken:

```python
dumpfn("Triclinic", "test.json")
```

The goal of `ValueEnum` is to define a set of string literals in the form of an enum. When we serialise these in a task document, we want to only store the enum value. E.g.

```python
{
    "crystal_system": "Triclinic"
}
```

rather than

```python
{
    "crystal_system": {
        "@module": "XYZ",
        "@class": "ValueEnum",
        "value": "Triclinic",
    }
}
```

The latter makes it convoluted to query the task document easily.

Perhaps there is a better way to achieve this and I agree that the `as_dict` labelling could be confusing. However, the point is this code has been in production for several years and works absolutely fine. The suggestion of enabling `enum_values=True` is not a good fix in this case, since this will mean it isn't possible to store ANY enums in their dict form.

To summarise, the issue raised in #914 is not a bug it is the intended behaviour as per the docstring and existing usage. This change has downstream impacts in atomate2. Please can we revert it.
